### PR TITLE
Implements the INSERT INTO ... SELECT ... statement.

### PR DESCRIPTION
**Pull request by jianqiao**
Implements the `INSERT INTO ... SELECT ...` statement. 
This feature provides convenience for populating relations.

NOTE: The behavior may be undefined when the destination relation appear in the SELECT query, e.g.
```
INSERT INTO a
SELECT * FROM a;
```
We may have a separate PR that either handles or prohibits this scenario.

### CODE REVIEW

**Comment by zuyu** [comment]
@jianqiao The problematic query you gave should be reported as errors during the query optimization where the resolver figures out it is a correlated query. Let alone that the catalog would throw an exception for existing relation names during the query execution.

File: parser/ParseStatement.hpp [hunk 1]
```diff
@@ -562,6 +566,43 @@ class ParseStatementInsert : public ParseStatement {
     return relation_name_.get();
   }
 
+ private:
+  std::unique_ptr<ParseString> relation_name_;
```
Comment (zuyu) [inline parser/ParseStatement.hpp]: Why adding a `private` member in the base class? It is not visible to its derived classes.

File: query_optimizer/physical/InsertSelection.cpp [hunk 2]
```diff
@@ -0,0 +1,45 @@
+/**
+ *   Copyright 2016, Quickstep Research Group, Computer Sciences Department,
+ *   University of Wisconsin—Madison.
```
Comment (zuyu) [inline query_optimizer/physical/InsertSelection.cpp]: Add two more whitespace indents.

File: parser/ParseStatement.hpp [hunk 3]
```diff
@@ -519,33 +519,37 @@ class ParseStatementSelect : public ParseStatement {
...
 /**
  * @brief The parsed representation of an INSERT statement.
+ *
+ * This is an abstract class where each of its subclass represents a concrete
+ * type of insert operation.
  **/
 class ParseStatementInsert : public ParseStatement {
  public:
+  enum class InsertType {
+    kTuple,
```
Comment (zuyu) [inline parser/ParseStatement.hpp]: Change to `kTuple = 0,`.

File: parser/ParseStatement.hpp [hunk 4]
```diff
@@ -519,33 +519,37 @@ class ParseStatementSelect : public ParseStatement {
...
    * @param relation_name The name of the relation to insert into.
-   * @param literal_values A list of literal values (in attribute-definition
-   *        order) to insert into the specified relation as a new tuple.
-   *        Becomes owned by this ParseStatementInsert.
    **/
   ParseStatementInsert(const int line_number,
                        const int column_number,
-                       ParseString *relation_name,
-                       PtrList<ParseScalarLiteral> *literal_values)
+                       ParseString *relation_name)
```
Comment (zuyu) [inline parser/ParseStatement.hpp]: Mark `const`.

File: parser/ParseStatement.hpp [hunk 5]
```diff
@@ -590,10 +631,86 @@ class ParseStatementInsert : public ParseStatement {
   }
 
  private:
-  std::unique_ptr<ParseString> relation_name_;
   std::unique_ptr<PtrList<ParseScalarLiteral> > literal_values_;
 
-  DISALLOW_COPY_AND_ASSIGN(ParseStatementInsert);
+  DISALLOW_COPY_AND_ASSIGN(ParseStatementInsertTuple);
+};
+
```
Comment (zuyu) [inline parser/ParseStatement.hpp]: Add one more empty line to separate two classes.

File: parser/ParseStatement.hpp (hunk 3 above)
Comment (pateljm) [inline parser/ParseStatement.hpp]: Thanks @zuyu -- to add some commentary, this is the style guide convention.

File: parser/ParseStatement.hpp (hunk 1 above)
Comment (pateljm) [inline parser/ParseStatement.hpp]: I think we can skip this comment and keep `relation_name_` private. @zuyu mentioned that he agrees with this view. Thanks!

**Comment by pateljm** [comment]
I think we are ready to close this PR. @zuyu -- do you agree?

**Comment by jianqiao** [comment]
@pateljm I can address the code style comments by @zuyu shortly after sending out a PR about the distinct operator tonight. Then we may close this PR. For the "recursive insert" issue mentioned at the top, I will create a new PR to address it.

File: parser/ParseStatement.hpp (hunk 5 above)
Comment (pateljm) [inline parser/ParseStatement.hpp]: Agreed with @jianqiao -- after this closes, lets do a separate PR for the two-line style.

Comment (zuyu) [inline parser/ParseStatement.hpp]: Sure.

File: query_optimizer/ExecutionGenerator.cpp [hunk 6]
```diff
@@ -1020,6 +1024,90 @@ void ExecutionGenerator::convertInsertTuple(
...
 }
 
+void ExecutionGenerator::convertInsertSelection(
+    const P::InsertSelectionPtr &physical_plan) {
+  // InsertSelection is converted to a Select and a SaveBlocks.
+
+  const CatalogRelationInfo *destination_relation_info =
+      findRelationInfoOutputByPhysical(physical_plan->destination());
+  const CatalogRelation &destination_relation =
+      *optimizer_context_->catalog_database()->getRelationByIdMutable(
```
Comment (zuyu) [inline query_optimizer/ExecutionGenerator.cpp]: Replace the last function with `getRelationById`.

File: query_optimizer/ExecutionGenerator.cpp [hunk 7]
```diff
@@ -1020,6 +1024,90 @@ void ExecutionGenerator::convertInsertTuple(
...
+          destination_relation_info->relation->getID());
```
Comment (zuyu) [inline query_optimizer/ExecutionGenerator.cpp]: Actually, refactor this statement to:

```
  const CatalogRelation &destination_relation = *destination_relation_info->relation;
```

File: query_optimizer/ExecutionGenerator.cpp [hunk 8]
```diff
@@ -1020,6 +1024,90 @@ void ExecutionGenerator::convertInsertTuple(
...
+  S::InsertDestination *insert_destination_proto = query_context_proto_->add_insert_destinations();
+
+  insert_destination_proto->set_insert_destination_type(S::InsertDestinationType::BLOCK_POOL);
+  insert_destination_proto->set_relation_id(destination_relation.getID());
+
+  const vector<block_id> blocks(destination_relation.getBlocksSnapshot());
+  for (const block_id block : blocks) {
+    insert_destination_proto->AddExtension(S::BlockPoolInsertDestination::blocks, block);
+  }
+
```
Comment (zuyu) [inline query_optimizer/ExecutionGenerator.cpp]: Need to set the storage block layout in `insert_destination_proto` to use `destination_relation`'s default one.

Refer to #104 for more info.

File: query_optimizer/ExecutionGenerator.cpp [hunk 9]
```diff
@@ -1020,6 +1024,90 @@ void ExecutionGenerator::convertInsertTuple(
...
+  const CatalogRelationInfo *selection_relation_info =
+      findRelationInfoOutputByPhysical(physical_plan->selection());
+  CatalogRelation *selection_relation =
+      optimizer_context_->catalog_database()->getRelationByIdMutable(
+          selection_relation_info->relation->getID());
```
Comment (zuyu) [inline query_optimizer/ExecutionGenerator.cpp]: Delete this one, and replace `selection_relation` in Line 1087 with `*selection_relation_info->relation`.

File: query_optimizer/ExecutionGenerator.cpp [hunk 10]
```diff
@@ -1020,6 +1024,90 @@ void ExecutionGenerator::convertInsertTuple(
...
+
+  // Prepare the attributes, which are output columns of the selection relation.
+  std::unique_ptr<std::vector<attribute_id>> attributes(new std::vector<attribute_id>());
+  for (E::AttributeReferencePtr attr_ref : physical_plan->selection()->getOutputAttributes()) {
+    unique_ptr<const Scalar> attribute(attr_ref->concretize(attribute_substitution_map_));
+
+    DCHECK_EQ(attribute->getDataSource(), Scalar::kAttribute);
```
Comment (zuyu) [inline query_optimizer/ExecutionGenerator.cpp]: Swap two arguments, as the rule is that the left hand side is the expected value, and the right the checked.

File: query_optimizer/resolver/Resolver.cpp [hunk 11]
```diff
@@ -687,8 +699,82 @@ L::LogicalPtr Resolver::resolveDropTable(
...
+    const Type& destination_type = destination_attributes[aid]->getValueType();
+    const Type& selection_type = selection_attributes[aid]->getValueType();
+    if (destination_type.equals(selection_type)) {
+      cast_expressions.emplace_back(selection_attributes[aid]);
+    } else {
+      if (destination_type.getSuperTypeID() == Type::kNumeric
+          && selection_type.getSuperTypeID() == Type::kNumeric
+          && destination_type.isSafelyCoercibleFrom(selection_type)) {
+        // Add cast operation
+        // TODO(jianqiao): implement Cast operation for non-numeric types.
```
Comment (zuyu) [inline query_optimizer/resolver/Resolver.cpp]: This `TODO` comment should go in the `else` part, as this `if` part is only about numeric types.

**Comment by zuyu** [comment]
@jianqiao LGTM, except for minor issues.

File: query_optimizer/ExecutionGenerator.cpp (hunk 8 above)
Comment (jianqiao) [inline query_optimizer/ExecutionGenerator.cpp]: Updated according to #104

File: query_optimizer/ExecutionGenerator.cpp (hunk 9 above)
Comment (jianqiao) [inline query_optimizer/ExecutionGenerator.cpp]: Updated.

File: query_optimizer/ExecutionGenerator.cpp (hunk 10 above)
Comment (jianqiao) [inline query_optimizer/ExecutionGenerator.cpp]: Updated.

File: query_optimizer/resolver/Resolver.cpp (hunk 11 above)
Comment (jianqiao) [inline query_optimizer/resolver/Resolver.cpp]: The original idea was that we can remove the
```
destination_type.getSuperTypeID() == Type::kNumeric
&& selection_type.getSuperTypeID() == Type::kNumeric
```
check once we extend `Cast`. So I just moved `TODO` above `if`.

**Comment by pateljm** [comment]
Looks good to merge. @jianqiao can you resync with master (the preprocessed files) and I'll close this PR.

**Comment by zuyu** [comment]
LGTM.

**Comment by zuyu** [comment]
@jianqiao Please remove the following lines in the commit message:

```
Code style fix for implementing the INSERT SELECT statement.

Minor fix for INSERT SELECT feature
```

When you squash commits, you should also re-word the commit message. Thanks!
